drivers/crypto/crypto_npcx_sha.c

```diff
@@ -0,0 +1,220 @@
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sha_npcx, CONFIG_CRYPTO_LOG_LEVEL);
+
+#define NPCX_HASH_CAPS_SUPPORT	(CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS)
+#define NPCX_SHA256_HANDLE_SIZE 212
+#define NPCX_SHA_MAX_SESSION	1
+
+/* The status code returns from Nuvoton Cryptographic Library ROM APIs */
+enum ncl_status {
+	NCL_STATUS_OK = 0xA5A5,
+	NCL_STATUS_FAIL = 0x5A5A,
+	NCL_STATUS_INVALID_PARAM = 0x02,
+	NCL_STATUS_PARAM_NOT_SUPPORTED,
+	NCL_STATUS_SYSTEM_BUSY,
+	NCL_STATUS_AUTHENTICATION_FAIL,
+	NCL_STATUS_NO_RESPONSE,
+	NCL_STATUS_HARDWARE_ERROR,
+};
+enum ncl_sha_type {
+	NCL_SHA_TYPE_2_256 = 0,
+	NCL_SHA_TYPE_2_384 = 1,
+	NCL_SHA_TYPE_2_512 = 2,
+	NCL_SHA_TYPE_NUM
+};
+
+/* The following table holds the function pointer for each SHA API in NPCX ROM. */
+struct npcx_ncl_sha {
+	/* Get the SHA context size required by SHA APIs. */
+	uint32_t (*get_context_size)(void);
+	/* Initial SHA context. */
+	enum ncl_status (*init_context)(void *ctx);
+	/* Finalize SHA context. */
+	enum ncl_status (*finalize_context)(void *ctx);
+	/* Initiate the SHA hardware module and setups needed parameters. */
+	enum ncl_status (*init)(void *ctx);
+	/*
+	 * Prepare the context buffer for a SHA calculation -  by loading the
+	 * initial SHA-256/384/512 parameters.
+	 */
+	enum ncl_status (*start)(void *ctx, enum ncl_sha_type type);
+	/*
+	 * Updates the SHA calculation with the additional data. When the
+	 * function returns, the hardware and memory buffer shall be ready to
+	 * accept new data * buffers for SHA calculation and changes to the data
+	 * in data buffer should no longer effect the SHA calculation.
+	 */
+	enum ncl_status (*update)(void *ctx, const uint8_t *data, uint32_t Len);
+	/* Return the SHA result (digest.) */
+	enum ncl_status (*finish)(void *ctx, uint8_t *hashDigest);
+	/* Perform a complete SHA calculation */
+	enum ncl_status (*calc)(void *ctx, enum ncl_sha_type type, const uint8_t *data,
+				uint32_t Len, uint8_t *hashDigest);
+	/* Power on/off the SHA module. */
+	enum ncl_status (*power)(void *ctx, uint8_t enable);
+	/* Reset the SHA hardware and terminate any in-progress operations. */
+	enum ncl_status (*reset)(void *ctx);
+};
+/* The start address of the SHA API table. */
+#define NPCX_NCL_SHA ((const struct npcx_ncl_sha *)DT_INST_REG_ADDR(0))
+
+struct npcx_sha_context {
+	uint8_t handle[NPCX_SHA256_HANDLE_SIZE];
+} __aligned(4);
+
+struct npcx_sha_session {
+	struct npcx_sha_context npcx_sha_ctx;
+	enum hash_algo algo;
+	bool in_use;
+};
+
+struct npcx_sha_session npcx_sessions[NPCX_SHA_MAX_SESSION];
+
+static int npcx_get_unused_session_index(void)
+{
+	int i;
+
+	for (i = 0; i < NPCX_SHA_MAX_SESSION; i++) {
+		if (!npcx_sessions[i].in_use) {
+			npcx_sessions[i].in_use = true;
+			return i;
+		}
+	}
+
+	return -1;
+}
+static int npcx_sha_compute(struct hash_ctx *ctx, struct hash_pkt *pkt, bool finish)
+{
+	enum ncl_status ret;
+	struct npcx_sha_session *npcx_session = (struct npcx_sha_session *)ctx->drv_sessn_state;
```

Comment:
`drv_sessn_state` is a `void *`, the casting here should not be needed. (same on other functions)

Reply:
Done.